python/ray/_private/worker.py

```diff
@@ -786,10 +786,30 @@ def get_objects(self, object_refs: list, timeout: Optional[float] = None):
 
     def main_loop(self):
         """The main loop a worker runs to receive and execute tasks."""
+        ray_terminate_msg = "SIGTERM is received to the process."
 
         def sigterm_handler(signum, frame):
-            shutdown(True)
-            sys.exit(1)
+            def exit_with_sigterm_error_message():
+                e = SystemExit(0)
```

Comment:
How much does this change the existing behaviour?

Can you maybe also help me understand how things have changed with this change?

Reply:
This is a bit hard to explain because the code is terrible. I will write an existing path in the future and propose the clean up https://docs.google.com/document/d/1mr4j5p9a0Ce5oYaUDoaiHUdOsbYfBOmAO4CPQTSjwR4/edit#heading=h.fwxjsbhjhpi5. 

Calling shutdown here is basically not a great idea because shutdown deallocates core_worker, and we should not do it until core worker finishes clean up. Here, calling sys.exit during task execution will trigger this code https://github.com/ray-project/ray/blob/f097cd45129ec6652ad84d9b3ecb82cb031755e6/python/ray/_raylet.pyx#L2057 (it is Python semantic. sys.exit == raising SystemExit), meaning the core worker is still used after sys.exit. shutdown is anyway called by atexit handler (which is called when interpreter exits), and I think that's the right place to call this because when an interpreter exit, core worker should've finished all cleanup. 

I think it worked so far because of mixed reasons, but basically it was just lucky it worked. 

The main difference is as follow;

Previous:
- If worker is idle: SIGTERM is just ignored (it is because python signal cannot be caught while running cpp code unless you call PyErr_CheckSignals). If it is killed by raylet, raylet sends sigkill in 100ms, and that kills the worker. 
- If worker is executing, shutdown() will clean up python side metadata and dealloc core worker. And then sys.exit will raise SystemExit to the execution code and core worker starts cleaning up and shutdown. 

After:
- If nothing is running, call Exit (same behavior as regular core worker shutdown path). 
- If anything is running, it raises SystemExit (which is basically the same behavior as sys.exit, same as status quo). But it doesn't call shutdown()